csrc/host_ir/jit.cpp

```diff
@@ -0,0 +1,383 @@
+  for (auto expr : container->topLevelExprs()) {
+    if (auto allocate = dynamic_cast<const kir::Allocate*>(expr)) {
+      // Generate a unique function name for this allocate
+      generateAllocateFunc(allocate, mod.get(), params);
+      // Store the mapping from allocate to function name
+      allocate_func_names[allocate] = "create_tensor_from_sizes_" +
+          std::to_string(reinterpret_cast<uintptr_t>(allocate));
+    }
+  }
+
+  // Add the module to the JIT
+  ExitOnErr(jit->addIRModule(
+      llvm::orc::ThreadSafeModule(std::move(mod), std::move(ctx))));
+
+  // Look up all functions and store their pointers
+  for (const auto& [allocate, func_name] : allocate_func_names) {
+    auto func_addr = ExitOnErr(jit->lookup(func_name));
+    // Lookup and reinterpret the function pointer to store in the map
+    allocate_funcs_[allocate] =
+        (allocate_fn)reinterpret_cast<at::Tensor* (*)(
+            const int64_t*, int64_t, const int64_t*, int64_t)>(func_addr.getValue());
+  }
+}
+
+// Constructor implementation
+HostIrJit::HostIrJit(
+    hir::HostIrContainer* container,
+    Communicator* communicator,
+    int num_threads)
+    : HostIrJit(
+          container,
+          communicator,
+          hir::HostIrEvaluatorParams{},
+          num_threads) {}
+
+HostIrJit::HostIrJit(
+    hir::HostIrContainer* container,
+    Communicator* communicator,
+    const hir::HostIrEvaluatorParams& params,
+    int num_threads)
+    : pimpl_(new LlvmJitImpl) {
+  // Initialize params with passed parameters
+  pimpl_->params_ =
+      std::make_unique<HostIrJitParams>(container, communicator, params);
+
+  llvm::InitializeNativeTarget();
+  llvm::InitializeNativeTargetAsmPrinter();
+  pimpl_->jit = ExitOnErr(
+      llvm::orc::LLJITBuilder().setNumCompileThreads(num_threads).create());
+  llvm::orc::JITDylib& dest_dynamic_lib = pimpl_->jit->getMainJITDylib();
+  auto mangler = llvm::orc::MangleAndInterner(
+      dest_dynamic_lib.getExecutionSession(), pimpl_->jit->getDataLayout());
+  dest_dynamic_lib.addGenerator(
+      ExitOnErr(llvm::orc::DynamicLibrarySearchGenerator::GetForCurrentProcess(
+          pimpl_->jit->getDataLayout().getGlobalPrefix())));
+
+  // Create wrapper function pointer to at::native::empty_strided_cuda
+  void* empty_strided_cuda_func_ptr = reinterpret_cast<void*>(
+      +[](const int64_t* sizes,
+          int64_t ndim,
+          const int64_t* strides,
+          int64_t strides_ndim,
+          int32_t dtype,
+          int64_t device_index) -> at::Tensor* {
+        at::IntArrayRef aten_sizes(sizes, ndim);
+        at::IntArrayRef aten_strides(strides, strides_ndim);
+        // Use the type and device passed as parameters
+        at::ScalarType scalar_type = static_cast<at::ScalarType>(dtype);
+        at::Device device =
+            at::Device(at::kCUDA, static_cast<c10::DeviceIndex>(device_index));
+        return new at::Tensor(at::native::empty_strided_cuda(
```

Comment:
why are we wrapping it as an at::Tensor*?
Like Jingyue mentioned earlier, `new` looks scary.